### PR TITLE
support L1 distance algorithm in hnsw/hgraph flat

### PR DESCRIPTION
**Pull request by jingyueob**
## Description
This is a feature which adds support for the L1 distance metric in VSAG vector indexes.

## Changes
Add a new distance metric by introducing L1Space for L1 distance computation and mapping VIDA_L1 to DISTANCE.

### CODE REVIEW

File: src/simd/basic_func.cpp [hunk 1]
```diff
@@ -19,6 +19,12 @@
 
 namespace vsag {
 
+static DistanceFuncType
+GetL1Distance() {
```
Comment (LHT129) [inline src/simd/basic_func.cpp:23]: [suggestion] The `GetL1Distance()` function only returns `generic::L1Distance` without any SIMD dispatch. All other distance functions (`GetL2Sqr`, `GetInnerProduct`, etc.) check for AVX512/AVX2/AVX/SSE/NEON/SVE support and dispatch to the appropriate SIMD implementation. This means L1 distance will always use the slow generic scalar path, even on hardware with SIMD support.

If SIMD-accelerated L1 distance implementations are planned for a follow-up PR, consider adding a TODO comment here. Otherwise, this could become a significant performance bottleneck for L1-based indexes.

```cpp
static DistanceFuncType
GetL1Distance() {
    return generic::L1Distance;
}
```

**Review by LHT129 — COMMENTED** [review]
Code review summary:

This PR adds L1 distance metric support to VSAG, restricted to HNSW and HGraph indexes with float32 dtype and fp32 quantization only. The implementation is clean and consistent with existing code patterns.

**What was reviewed:**
- `space_l1.h`: New L1Space class for hnswlib, follows the same pattern as L2Space
- `basic_func.h/cpp`: L1Distance function pointer with dispatch (currently only generic path)
- `generic.cpp`: Generic scalar L1 distance implementation using `std::fabs`
- `metric_type.h`: Added `METRIC_TYPE_L1 = 3`
- `flatten_interface.cpp`: L1-specific validation and instantiation path
- `engine.cpp`: Index-level restriction to HNSW and HGraph only
- `index_common_param.cpp`: Metric type parsing and dtype validation
- `hnsw_zparameters.cpp`: L1Space instantiation in HNSW parameter creation
- `fp32_quantizer.cpp`: L1 distance computation in ComputeImpl, ComputeDistImpl, and ComputeDistsBatch4Impl
- `eval_dataset.cpp`: L1/manhattan distance support in evaluation tools
- Tests: Comprehensive unit tests and functional tests added across hnsw_test.cpp, hnsw_zparameters_test.cpp, factory_test.cpp, fp32_quantizer_test.cpp, basic_func_test.cpp, index_commom_param_test.cpp, test_hgraph.cpp, test_hnsw.cpp, and test_dataset.cpp

**Overall assessment:** The implementation is correct, well-tested, and follows the existing codebase conventions. The restrictions (float32 only, fp32 quantization only, HNSW/HGraph only) are properly validated at multiple layers. No blocking issues found.

File: src/algorithm/hnswlib/space_l1.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    void*
+    get_dist_func_param() override {
+        return &dim_;
+    }
+
+private:
+    DISTFUNC fstdistfunc_;
+    size_t data_size_;
+    uint64_t dim_;
```
Comment (LHT129) [inline src/algorithm/hnswlib/space_l1.h:46]: [note] `dim_` is declared as `uint64_t` but `data_size_` is declared as `size_t`. This is inconsistent within the same class. The VSAG coding standards prefer `uint64_t` over `size_t` to avoid macOS compilation issues. Consider using `uint64_t` for `data_size_` as well (the `dim * sizeof(float)` computation already involves `uint64_t` from `dim_`).

Note that `L2Space` uses `size_t` for both fields, so either approach is acceptable as long as it is consistent within the class.

File: src/algorithm/hnswlib/space_l1.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    get_dist_func() override {
+        return fstdistfunc_;
+    }
+
+    void*
+    get_dist_func_param() override {
+        return &dim_;
+    }
+
+private:
```
Comment (LHT129) [inline src/algorithm/hnswlib/space_l1.h:43]: [suggestion] `data_size_` is declared as `size_t` but the VSAG coding standard recommends using `uint64_t` instead of `size_t` to avoid macOS compilation issues. The constructor already uses `uint64_t dim` and computes `dim * sizeof(float)`, so the result should be stored as `uint64_t` for consistency.

```cpp
// current:
size_t data_size_;

// suggested:
uint64_t data_size_;
```

Also, `get_data_size()` returns `size_t` per the `SpaceInterface` override — if the interface requires `size_t`, consider adding a `static_cast` or changing the interface separately.

File: src/quantization/fp32_quantizer.cpp [hunk 4]
```diff
@@ -245,6 +253,12 @@ FP32Quantizer<metric>::ComputeDistsBatch4Impl(Computer<FP32Quantizer<metric>>& c
                                dists2,
                                dists3,
                                dists4);
+    } else if constexpr (metric == MetricType::METRIC_TYPE_L1) {
```
Comment (LHT129) [inline src/quantization/fp32_quantizer.cpp:256]: [suggestion] `ComputeDistsBatch4Impl` for `METRIC_TYPE_L1` calls `L1Distance` four separate times, each iterating over the full dimension. For other metrics like `METRIC_TYPE_L2SQR`, the batch4 path uses a fused `FP32ComputeL2SqrBatch4` that computes all four distances in a single pass over the dimension.

Consider implementing a fused `L1DistanceBatch4` that computes all four L1 distances in one loop to reduce memory bandwidth pressure and improve cache efficiency. This is especially relevant since `ComputeDistsBatch4Impl` is explicitly designed for batched computation.

File: src/quantization/fp32_quantizer.cpp [hunk 5]
```diff
@@ -245,6 +253,12 @@ FP32Quantizer<metric>::ComputeDistsBatch4Impl(Computer<FP32Quantizer<metric>>& c
...
+        const uint64_t dim = this->dim_;
+        dists1 = L1Distance(computer.buf_, codes1, &dim);
+        dists2 = L1Distance(computer.buf_, codes2, &dim);
```
Comment (LHT129) [inline src/quantization/fp32_quantizer.cpp]: [suggestion] The `ComputeDistsBatch4Impl` L1 branch computes four distances with four separate `L1Distance` calls, while `L2SQR`, `IP`, and `COSINE` all use dedicated batch-4 SIMD functions (`FP32ComputeL2SqrBatch4`, `FP32ComputeIPBatch4`). For consistency and potential performance, consider adding an `L1DistanceBatch4` SIMD function or at least noting this as a future optimization.

```cpp
// current L1 branch in ComputeDistsBatch4Impl:
const uint64_t dim = this->dim_;
dists1 = L1Distance(computer.buf_, codes1, &dim);
dists2 = L1Distance(computer.buf_, codes2, &dim);
dists3 = L1Distance(computer.buf_, codes3, &dim);
dists4 = L1Distance(computer.buf_, codes4, &dim);
```

File: src/simd/basic_func.cpp (hunk 1 above)
Comment (jingyueob) [reply src/simd/basic_func.cpp:23]: done

File: tools/eval/eval_dataset.cpp [hunk 6]
```diff
@@ -261,6 +261,8 @@ EvalDataset::Load(const std::string& filename) {
                                    std::sqrt(vsag::InnerProduct(query1, query1, qty_ptr) *
                                              vsag::InnerProduct(query2, query2, qty_ptr));
                 };
+            } else if (metric == "manhattan" || metric == "l1") {
```
Comment (LHT129) [inline tools/eval/eval_dataset.cpp]: [note] The eval tool in `tools/eval/eval_dataset.cpp` accepts `"manhattan"` as an alias for L1 distance, but `IndexCommonParam::CheckAndCreate` in `src/index_common_param.cpp` only accepts `"l1"` (lowercase). The test in `src/index_commom_param_test.cpp` explicitly verifies that aliases like `"manhattan"` and `"L1"` are rejected.

This inconsistency means a dataset with `metric="manhattan"` can be loaded by the eval tool but the same metric name cannot be used to create an index. Consider either:
- Adding `"manhattan"` as an accepted alias in `fill_metrictype`, or
- Removing the `"manhattan"` alias from the eval tool for consistency.

```cpp
// eval_dataset.cpp (accepts manhattan):
} else if (metric == "manhattan" || metric == "l1") {

// index_common_param.cpp (only accepts l1):
} else if (metric == METRIC_L1) {  // METRIC_L1 = "l1"
```

File: src/quantization/fp32_quantizer.cpp (hunk 4 above)
Comment (jingyueob) [reply src/quantization/fp32_quantizer.cpp:256]: currently I only implements generic version

**Review by LHT129 — COMMENTED** [review]
The incremental changes in this push address all the previously raised review comments:

1. **`src/algorithm/hnswlib/space_l1.h`**: Changed `data_size_` from `size_t` to `uint64_t` for consistency with VSAG coding standards.

2. **`src/simd/basic_func.cpp`**: Added TODO comment noting that SIMD-accelerated L1 implementations are planned for a follow-up PR.

3. **`src/quantization/fp32_quantizer.cpp`**: Replaced four separate `L1Distance` calls in `ComputeDistsBatch4Impl` with a single fused `L1DistanceBatch4` call, matching the pattern used by other metrics (`FP32ComputeL2SqrBatch4`, `FP32ComputeIPBatch4`).

4. **`src/simd/fp32_simd.cpp` / `fp32_simd.h`**: Added `L1DistanceBatch4` dispatch function with TODO for future SIMD implementations.

5. **`src/simd/generic.cpp`**: Added generic `L1DistanceBatch4` implementation that computes all four distances in a single pass over the dimension.

6. **`src/simd/fp32_simd_test.cpp`**: Added test validating that `L1DistanceBatch4` results match standalone `L1Distance` calls across multiple dimensions.

7. **`tools/eval/eval_dataset.cpp`**: Removed the `"manhattan"` alias, keeping only `"l1"` for consistency with `IndexCommonParam::CheckAndCreate`.

The new `L1DistanceBatch4` implementation follows the existing codebase patterns precisely, and the test coverage is adequate. No new issues found.

File: tools/eval/eval_dataset.cpp [hunk 7]
```diff
@@ -261,6 +261,8 @@ EvalDataset::Load(const std::string& filename) {
                                    std::sqrt(vsag::InnerProduct(query1, query1, qty_ptr) *
                                              vsag::InnerProduct(query2, query2, qty_ptr));
                 };
+            } else if (metric == "l1") {
```
Comment (LHT129) [inline tools/eval/eval_dataset.cpp:264]: [note] This change correctly removes the `"manhattan"` alias to be consistent with `IndexCommonParam::CheckAndCreate` which only accepts `"l1"`. The test in `src/index_commom_param_test.cpp` explicitly verifies that aliases like `"manhattan"` are rejected, so this alignment is the right approach.

**Review by LHT129 — COMMENTED** [review]
Overall this PR is well-implemented. The L1 distance metric is correctly integrated across the full stack: SIMD primitives, hnswlib space, quantization, factory/index creation, and eval tooling. Test coverage is thorough with unit tests for each layer and functional tests for both HNSW and HGraph.

A few observations:

1. **`src/algorithm/hnswlib/space_l1.h:46`** — `data_size_` is declared as `size_t` while `dim_` is `uint64_t`. This inconsistency within the same class was noted in a prior review and still remains. Since `dim_` is `uint64_t` and `dim * sizeof(float)` produces a `uint64_t` value, `data_size_` should also be `uint64_t` for consistency.

2. **`src/datacell/flatten_interface.cpp:159-161`** — The dtype validation (`l1 metric must use float32 dtype`) is duplicated from `src/index_common_param.cpp:117-119`. While defense-in-depth is reasonable, consider whether the duplicate check adds value or just maintenance burden.

3. **`src/index/hnsw_zparameters.cpp:30-32`** — Similarly, the L1 dtype check here duplicates the validation already performed in `IndexCommonParam::CheckAndCreate`. By the time `HnswParameters::FromJson` is called, the dtype has already been validated.

The SIMD TODO comments in `basic_func.cpp` and `fp32_simd.cpp` correctly indicate that SIMD-accelerated L1 implementations are planned for a follow-up PR. The generic scalar implementations are correct and sufficient for an initial feature landing.

**Review by LHT129 — COMMENTED** [review]
Overall this PR is well-implemented. The L1 distance metric is correctly integrated across the full stack: SIMD primitives, hnswlib space, quantization, factory/index creation, and eval tooling. Test coverage is thorough with unit tests for each layer and functional tests for both HNSW and HGraph.

A few observations:

1. **`src/algorithm/hnswlib/space_l1.h:46`** — `data_size_` is declared as `size_t` while `dim_` is `uint64_t`. This inconsistency within the same class was noted in a prior review and still remains. Since `dim_` is `uint64_t` and `dim * sizeof(float)` produces a `uint64_t` value, `data_size_` should also be `uint64_t` for consistency.

2. **`src/datacell/flatten_interface.cpp:159-161`** — The dtype validation (`l1 metric must use float32 dtype`) is duplicated from `src/index_common_param.cpp:117-119`. While defense-in-depth is reasonable, consider whether the duplicate check adds value or just maintenance burden.

3. **`src/index/hnsw_zparameters.cpp:30-32`** — Similarly, the L1 dtype check here duplicates the validation already performed in `IndexCommonParam::CheckAndCreate`. By the time `HnswParameters::FromJson` is called, the dtype has already been validated.

The SIMD TODO comments in `basic_func.cpp` and `fp32_simd.cpp` correctly indicate that SIMD-accelerated L1 implementations are planned for a follow-up PR. The generic scalar implementations are correct and sufficient for an initial feature landing.

**Review by LHT129 — COMMENTED** [review]
This latest commit (7d38d64) refactors the L1 metric validation from `CHECK_ARGUMENT` macro to explicit `if` + `throw VsagException` in two files (`src/index_common_param.cpp` and `src/index/hnsw_zparameters.cpp`). The change is correct and improves readability by making the validation logic more explicit.

The overall PR adds L1 distance support to HNSW/HGraph indexes with:
- New `L1Space` class for hnswlib integration
- `METRIC_TYPE_L1` enum value and `"l1"` metric string constant
- Generic scalar `L1Distance` implementation (with TODO for future SIMD)
- `L1DistanceBatch4` for batched computation
- Proper validation: L1 only supports float32 dtype, fp32 quantization, flatten data cell, and HNSW/HGraph indexes
- Comprehensive test coverage across unit and functional tests

One minor note: `data_size_` in `L1Space` uses `size_t` while `dim_` uses `uint64_t` — consider using `uint64_t` for consistency per VSAG coding standards, though this is not a blocker since `L2Space` follows the same pattern.

File: src/datacell/flatten_interface.cpp [hunk 8]
```diff
@@ -156,6 +156,21 @@ template <typename IOTemp>
...
+        }
+        if (param->quantizer_parameter->GetTypeName() != QUANTIZATION_TYPE_VALUE_FP32) {
+            throw VsagException(ErrorType::INVALID_ARGUMENT,
+                                "l1 metric supports only fp32 quantization");
+        }
+        return make_instance_flatten<FP32Quantizer<MetricType::METRIC_TYPE_L1>, IOTemp>(
+            param, common_param);
+    }
     if (metric == MetricType::METRIC_TYPE_L2SQR) {
         return make_instance<MetricType::METRIC_TYPE_L2SQR, IOTemp>(param, common_param);
```
Comment (LHT129) [inline src/datacell/flatten_interface.cpp:175]: [note] The L1 metric path in `make_instance` bypasses the `make_instance<metric, IOTemp>` template dispatch and directly calls `make_instance_flatten<FP32Quantizer<METRIC_TYPE_L1>, IOTemp>`. This means L1 does not go through the `is_transform_quantizer` logic that handles TQ (TransformQuantizer) wrapping for other metrics.

If TQ support for L1 is not planned, this is fine. But if it might be added in the future, consider routing L1 through the same `make_instance<METRIC_TYPE_L1, IOTemp>` path and adding the necessary quantization-type restrictions there instead of the early-return checks here.

```cpp
// current: early return bypasses make_instance<metric, IOTemp>
if (metric == MetricType::METRIC_TYPE_L1) {
    // ... validation checks ...
    return make_instance_flatten<FP32Quantizer<MetricType::METRIC_TYPE_L1>, IOTemp>(param, common_param);
}
```

**Review by LHT129 — COMMENTED** [review]
Thanks for this PR adding L1 distance support. The latest commit (7d38d64) has addressed all the issues raised in the previous review round:

- `GetL1Distance()` now includes a TODO for future SIMD acceleration
- `space_l1.h` now uses `uint64_t` consistently for `data_size_`
- `ComputeDistsBatch4Impl` now uses the fused `L1DistanceBatch4` function
- The `"manhattan"` alias has been removed from `eval_dataset.cpp` for consistency

The implementation is clean and well-structured: L1 is properly restricted to hnsw/hgraph indexes with fp32 quantization only, and the test coverage is thorough across unit tests, functional tests, and the eval tool.

Left one minor [note] about the `flatten_interface.cpp` dispatch path, but it is not blocking.

**Review by LHT129 — COMMENTED** [review]
## Code Review Summary

This PR adds L1 (Manhattan) distance metric support to hnsw/hgraph flat indexes. Overall code quality is solid.

### What was reviewed
- 26 files changed, +580/-56 lines
- New `space_l1.h` for hnswlib L1 space
- New `L1Distance` and `L1DistanceBatch4` functions in simd layer (generic implementation, SIMD TODO noted for follow-up)
- New `METRIC_TYPE_L1 = 3` enum value and `METRIC_L1 = "l1"` constant
- L1 support in index_common_param, hnsw_zparameters, engine, flatten_interface, fp32_quantizer
- Comprehensive test coverage across unit tests, functional tests, and ground truth computation

### Positive observations
- Previous review comments from LHT129 have been addressed: `GetL1Distance()` now has a TODO for SIMD acceleration, and `data_size_` is correctly typed as `uint64_t`
- Validation is thorough at multiple layers: dtype must be float32, only hnsw/hgraph indexes, only flatten data cell, only fp32 quantization
- Test coverage is comprehensive: factory creation, index_common_param, hnsw_zparameters, hnsw unit tests, hgraph functional tests, basic_func, fp32_simd, fp32_quantizer, test_dataset ground truth, eval_dataset

### Issues found
No substantive issues identified. The implementation is clean, follows existing patterns, and all edge cases appear properly handled.

**Review by wxyucs — APPROVED** [review]
lgtm

**Review by LHT129 — APPROVED** [review]
LGTM
